 ## <span style='background :yellow' > Question 1: What percentage of trades are with 30+ DPD (more than 30 days past due) among all the trades available? </span>

In [1]:
!pip install lxml

You should consider upgrading via the 'C:\Users\Rutik\AppData\Local\Programs\Python\Python310\python.exe -m pip install --upgrade pip' command.


In [2]:
# Importing libraries 
import os
import json
from bs4 import BeautifulSoup
import pandas as pd

In [3]:
os.listdir()[1:11]

['customer1113697_loan1115483_crif_report.html.xml',
 'customer1129550_loan1131339_crif_report.html.xml',
 'customer1195586_loan1197471_crif_report.html.xml',
 'customer14235_loan14235_crif_report.html.xml',
 'customer16475_loan16475_crif_report.html.xml',
 'customer40409_loan40409_crif_report.html.xml',
 'customer773504_loan774538_crif_report.html.xml',
 'customer787561_loan788638_crif_report.html.xml',
 'customer794397_loan795497_crif_report.html.xml',
 'customer898231_loan899591_crif_report.html.xml']

In [4]:


history=[]

for i in os.listdir()[1:11]:  
    with open(i, 'r') as f:   # Opening all 10 customer files
        xml_data=f.read()
        bs_data = BeautifulSoup(xml_data, 'xml') 
        payment_history=bs_data.find_all("COMBINED-PAYMENT-HISTORY")  # Getting Payment data From all XML files
        for i in payment_history:
                history.append(i.text)
                

History_split=[]
import re

# Removing Unwanted Object from data
for i in history:

    a=re.split(r'[,|,\s,'',]',i)
#     print(a,'\n')
    History_split.append(a)

        
for i in History_split:       # Removing unwanted using pop 
    removespace=i.pop(-1)
    

    
# Coverting Data Into dictionary (key:Value Format)
main_history=[]     
def Convert(lst):
    res_dct = map(lambda i: (lst[i], lst[i+1]), range(len(lst)-1)[::2])
    return dict(res_dct)
 
for i in History_split:
    sample=(Convert(i))
    main_history.append(sample)
    
    
data=pd.DataFrame(main_history)             # Creating DataFrame

# Replacing unwanted part of data into emptystring
for i in data.columns:           
    data[i]=data[i].str.replace('/','').astype(str)

# Replacing unwanted part of data into emptystring    
for i in data.columns:
    data[i]=(data[i].str.replace("[XXXSTDDDDSUBDBTLO]",""))
 

#removing nan string data into empty string
for i in data.columns:
    data[i]=(data[i].str.replace("[nan]",""))
    
    
list1=data.values.tolist()  # Converting list data into DataFrame with Pandas


NewList=[]                # Seperating Numeric Data For Calculation
for a in list1:
     for b in a:
          if b.isnumeric():
                NewList.append(b)
            
NewList=list(map(int,NewList))     # Converting Datatype String into integer

C:\Users\Rutik\AppData\Local\Temp/ipykernel_23544/1333465780.py:47: FutureWarning: The default value of regex will change from True to False in a future version.
  data[i]=(data[i].str.replace("[XXXSTDDDDSUBDBTLO]",""))
C:\Users\Rutik\AppData\Local\Temp/ipykernel_23544/1333465780.py:52: FutureWarning: The default value of regex will change from True to False in a future version.
  data[i]=(data[i].str.replace("[nan]",""))


In [5]:
df=pd.DataFrame(NewList) # CReating DataFrame
df.rename(columns={0: "Days_Past_Due"}, inplace=True) # Changing Column name in the DataFrame


In [6]:
df       # Data of past dues

Days_Past_Due
0                 0
1                 0
2                 0
3               175
4               146
...             ...
2577              0
2578              0
2579              0
2580              0
2581              0

[2582 rows x 1 columns]

In [7]:
# Calculating data which have  30+ DPD
trades_with_30Plus =0    
for i in NewList:
    if i >30:
        trades_with_30Plus+=1
trades_with_30Plus

391

In [8]:
Answer=( trades_with_30Plus/len(NewList)*100)
print('trades are with 30+ DPD (more than 30 days past due) among all  the trades available',Answer,"%")

trades are with 30+ DPD (more than 30 days past due) among all  the trades available 15.143299767621999 %


### Answer: Question 1
##  Percentage Trades are with 30+ DPD (more than 30 days past due) among all  the trades available = 15.143299767621999 %

*** 
*** 
*** 

## <span style='background :yellow' > Question 2: What is the sum of total disbursed amount for all loans for each customer? </span>

In [9]:
# Importing imp libraries 
import os
import json
from bs4 import BeautifulSoup
import pandas as pd

In [10]:
os.listdir()[1:11]

['customer1113697_loan1115483_crif_report.html.xml',
 'customer1129550_loan1131339_crif_report.html.xml',
 'customer1195586_loan1197471_crif_report.html.xml',
 'customer14235_loan14235_crif_report.html.xml',
 'customer16475_loan16475_crif_report.html.xml',
 'customer40409_loan40409_crif_report.html.xml',
 'customer773504_loan774538_crif_report.html.xml',
 'customer787561_loan788638_crif_report.html.xml',
 'customer794397_loan795497_crif_report.html.xml',
 'customer898231_loan899591_crif_report.html.xml']

In [11]:
Loan_amount=[]
history=[]

# Reading all XML files of Customer
for i in os.listdir()[1:-3]:       
    with open(i, 'rb') as f:
        xml_data=f.read()
        bs_data = BeautifulSoup(xml_data, 'xml') 
        
        DISBURSED_AMT=bs_data.find_all("DISBURSED-AMT")     # Getting Data
        
        for i in DISBURSED_AMT:
                Loan_amount.append(i.text)                  # Text data of DISBURSED AMT
                
    



In [12]:
Final_Amount=[]
for i in Loan_amount:                                      # Splitting Data and appending into New List   
        b=''.join(str(i).split(','))
        Final_Amount.append(b)



Final_Amount                                               # Calculating Sum of Data

['35000',
 '31500',
 '31500',
 '12000',
 '6400',
 '15500',
 '70000',
 '45000',
 '20000',
 '47000',
 '35000',
 '20000',
 '5000',
 '20000',
 '130000',
 '21600',
 '21600',
 '50000',
 '20000',
 '165000',
 '11500',
 '11500',
 '40000',
 '46980',
 '11340',
 '38000',
 '52000',
 '62000',
 '35000',
 '12500',
 '10100',
 '22000',
 '22000',
 '9500',
 '32500',
 '8100',
 '1',
 '9200',
 '65100',
 '11100',
 '500000',
 '93500',
 '129000',
 '87700',
 '291500',
 '45614',
 '45614',
 '45614',
 '45614',
 '299000',
 '26740',
 '87700',
 '317000',
 '100000',
 '129000',
 '129000',
 '638536',
 '87700',
 '16032',
 '85841',
 '291500',
 '1710000',
 '1800000',
 '1710000',
 '824000',
 '100000',
 '127341',
 '89258',
 '900000',
 '130405',
 '22999',
 '375000',
 '35000',
 '121094',
 '200000',
 '300000',
 '42000',
 '60000',
 '70000',
 '301000',
 '985000',
 '57580',
 '1228599',
 '650000',
 '34819',
 '47148',
 '532',
 '55666',
 '100000',
 '375000',
 '130000',
 '130000',
 '97983',
 '114304',
 '35269',
 '62055',
 '34028',
 '10

In [13]:
df2=pd.DataFrame(Final_Amount)                             # Creating DataFrame 
df2.rename(columns={0: "DISBURSED-AMT"}, inplace=True)     # Renaming The column for Better Understanding


In [14]:
df2

DISBURSED-AMT
0           35000
1           31500
2           31500
3           12000
4            6400
..            ...
277         50300
278         21990
279         41700
280          7000
281         16600

[282 rows x 1 columns]

In [15]:
df2 = df2.astype({'DISBURSED-AMT':'int'})                  # Changing Datatype of data
df2.sum()  

DISBURSED-AMT    29573391
dtype: int64

In [16]:
df2

DISBURSED-AMT
0            35000
1            31500
2            31500
3            12000
4             6400
..             ...
277          50300
278          21990
279          41700
280           7000
281          16600

[282 rows x 1 columns]

In [17]:
df2.loc['Grand_total']= df2.sum()         # Adding Column Grand Total Into DataFrame
df2

DISBURSED-AMT
0                    35000
1                    31500
2                    31500
3                    12000
4                     6400
...                    ...
278                  21990
279                  41700
280                   7000
281                  16600
Grand_total       29573391

[283 rows x 1 columns]

### Answer: Question 2
## Sum of total disbursed amount for all loans for each customer :42579056

*** 
---
___


## <span style='background :yellow' > Question 3:What is the maximum number of months of 30+ due per trade was there?
 </span>

In [18]:
history=[]

for i in os.listdir()[1:11]:  
    with open(i, 'rb') as f:   # Opening all 10 customer files
        xml_data=f.read()
        bs_data = BeautifulSoup(xml_data, 'xml') 
        payment_history=bs_data.find_all("COMBINED-PAYMENT-HISTORY")  # Getting Payment data From all XML files
        for i in payment_history:
                history.append(i.text)
                

History_split=[]
import re

# Removing Unwanted Object from data
for i in history:

    a=re.split(r'[,|,\s,'',]',i)
#     print(a,'\n')
    History_split.append(a)

        
for i in History_split:       # Removing unwanted using pop 
    removespace=i.pop(-1)
    

    
# Coverting Data Into dictionary (key:Value Format)
main_history=[]     
def Convert(lst):
    res_dct = map(lambda i: (lst[i], lst[i+1]), range(len(lst)-1)[::2])
    return dict(res_dct)
 
for i in History_split:
    sample=(Convert(i))
    main_history.append(sample)
    
    
data1=pd.DataFrame(main_history)             # Creating DataFrame

# Replacing unwanted part of data into emptystring
for i in data.columns:           
    data1[i]=data1[i].str.replace('/','').astype(str)

# Replacing unwanted part of data into emptystring    
for i in data.columns:
    data1[i]=(data1[i].str.replace("[XXXSTDDDDSUBDBTLO]",""))
 

#removing nan string data into empty string
for i in data1.columns:
    data1[i]=(data1[i].str.replace("[nan]",""))
    
    
list1=data1.values.tolist()  # Converting list data into DataFrame with Pandas

C:\Users\Rutik\AppData\Local\Temp/ipykernel_23544/171038715.py:47: FutureWarning: The default value of regex will change from True to False in a future version.
  data1[i]=(data1[i].str.replace("[XXXSTDDDDSUBDBTLO]",""))
C:\Users\Rutik\AppData\Local\Temp/ipykernel_23544/171038715.py:52: FutureWarning: The default value of regex will change from True to False in a future version.
  data1[i]=(data1[i].str.replace("[nan]",""))


In [19]:
# Seprating Numeric values for calculation and adding them into new list

NewList=[]
for a in list1:
     x=[]
     for b in a:
          if b.isnumeric():
                x.append(b)
     NewList.append(x)

In [20]:
# Removing Empty list Data from DATA

res = [ele for ele in NewList if ele != []]

# Changing Data Type of the Data
new_l=[]
for i in res:
    h=list(map(int, i))
    new_l.append(h)
    
new_l


[[0, 0],
 [0, 175, 146, 0, 0, 108, 46, 1, 0, 0, 0],
 [114, 86, 0, 55, 0],
 [0, 0, 0, 0],
 [0, 0, 0, 0],
 [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
 [0],
 [0, 0, 0, 0, 0, 0, 0, 0, 0],
 [0],
 [0],
 [0, 0],
 [0],
 [0, 0, 0, 0, 0, 0, 0],
 [0, 172, 143, 0, 0, 108, 46, 1, 0, 0, 0],
 [0, 0],
 [0],
 [120, 0, 210, 180, 150],
 [73, 0, 0, 0, 0],
 [0, 0, 0, 0, 0],
 [0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0],
 [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
 [0, 0, 0, 0, 0, 0, 0, 0, 0],
 [0],
 [0],
 [0],
 [0],
 [113, 82, 0, 141],
 [0, 0, 0, 0, 0, 0, 0],
 [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
 [0, 0, 0],
 [0, 0, 0],
 [0, 53],
 [0, 93, 62, 0, 0, 0, 0, 32],
 [334, 304, 273, 242, 212, 181, 151, 120, 92, 1],
 [0, 0, 0],
 [0, 0],
 [0, 0, 28, 0, 0, 0],
 [0, 0, 0],
 [0, 0, 125, 94, 31, 2, 0, 0, 64],
 [0, 0, 0, 0, 90, 59, 28, 0, 0, 0],
 [0, 0, 0, 0, 90, 59, 28, 0, 0, 0],
 [80, 52, 0, 21, 0, 0, 0,

In [21]:
# Finding Values Which Are greater than 30
trade=[]
for i in new_l :
    k=[]
    for j in i:
        if j > 30:
            k.append(j)
    trade.append(k)
trade

[[],
 [175, 146, 108, 46],
 [114, 86, 55],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [172, 143, 108, 46],
 [],
 [],
 [120, 210, 180, 150],
 [73],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [113, 82, 141],
 [],
 [],
 [],
 [],
 [53],
 [93, 62, 32],
 [334, 304, 273, 242, 212, 181, 151, 120, 92],
 [],
 [],
 [],
 [],
 [125, 94, 31, 64],
 [90, 59],
 [90, 59],
 [80, 52],
 [],
 [66, 35],
 [92, 62, 31],
 [],
 [],
 [118, 88, 57],
 [],
 [],
 [],
 [],
 [68, 37],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [60],
 [],
 [90, 90, 90, 60, 120, 120, 90],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [56],
 [180, 150],
 [738],
 [],
 [],
 [],
 [60, 60],
 [],
 [],
 [120, 90, 60],
 [],
 [],
 [],
 [],
 [],
 [54, 56, 57, 86, 56],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [77, 46],
 [],
 [52],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [78, 47],
 [49],
 [],
 [55],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [39],
 [],
 [],
 [],
 [],
 [],
 [329, 298, 270, 239, 178],
 [58],
 [],
 [],
 [],
 [],


In [22]:
# Showing 30+ DPD occurance 

maximum=[]
Statement=[]
b=1
for i in trade:
        maximum.append(len(i))
        x=(' The number of occurrences which are 30+ DPD For Trade' ,b, 'is',len(i))
        Statement.append(x)
        
        b+=1

In [23]:
Statement

[(' The number of occurrences which are 30+ DPD For Trade', 1, 'is', 0),
 (' The number of occurrences which are 30+ DPD For Trade', 2, 'is', 4),
 (' The number of occurrences which are 30+ DPD For Trade', 3, 'is', 3),
 (' The number of occurrences which are 30+ DPD For Trade', 4, 'is', 0),
 (' The number of occurrences which are 30+ DPD For Trade', 5, 'is', 0),
 (' The number of occurrences which are 30+ DPD For Trade', 6, 'is', 0),
 (' The number of occurrences which are 30+ DPD For Trade', 7, 'is', 0),
 (' The number of occurrences which are 30+ DPD For Trade', 8, 'is', 0),
 (' The number of occurrences which are 30+ DPD For Trade', 9, 'is', 0),
 (' The number of occurrences which are 30+ DPD For Trade', 10, 'is', 0),
 (' The number of occurrences which are 30+ DPD For Trade', 11, 'is', 0),
 (' The number of occurrences which are 30+ DPD For Trade', 12, 'is', 0),
 (' The number of occurrences which are 30+ DPD For Trade', 13, 'is', 0),
 (' The number of occurrences which are 30+ DPD

In [24]:
df3=pd.DataFrame(Statement)

In [25]:
# Calculating Max number of moths with 30+ Occurance

In [26]:
print('The maximum number of months of 30+ due per trade:', max(maximum))

The maximum number of months of 30+ due per trade: 36


### Answer: Question 3
# The maximum number of months of 30+ due per trade is 36

*** 
---
___


## importing data of All Questions into Excel

In [27]:

with pd.ExcelWriter("zproject.xlsx", engine="openpyxl") as writer:
    df.to_excel(writer, sheet_name="Question_1",header=True, index=False)
    df2.to_excel(writer, sheet_name="Question_2",  header=True, index=False) 
    df3.to_excel(writer, sheet_name="Question_3", header=True, index=False)

ModuleNotFoundError: No module named 'openpyxl'